node/src/components/consensus/highway_core/finality_detector/rewards.rs

```diff
@@ -22,29 +23,20 @@ pub(crate) fn compute_rewards<C: Context>(state: &State<C>, bhash: &C::Hash) ->
     // The panorama of the payout block: Rewards must only use this panorama, since it defines
     // what everyone who has the block can already see.
     let panorama = &payout_vote.panorama;
-    // The vote that introduced the payout block's parent.
-    let opt_parent_vote = payout_block.parent().map(|h| state.vote(h));
-    // The parent's timestamp, or 0.
-    let parent_time = opt_parent_vote.map_or_else(Timestamp::zero, |vote| vote.timestamp);
-    // Only summits for blocks for which rewards are scheduled between the previous and current
-    // timestamps are rewarded, and only if they are ancestors of the payout block.
-    let range = parent_time..payout_vote.timestamp;
-    let is_ancestor =
-        |bh: &&C::Hash| Some(*bh) == state.find_ancestor(bhash, state.block(bh).height);
     let mut rewards = ValidatorMap::from(vec![0u64; panorama.len()]);
-    for bhash in state.rewards_range(range).filter(is_ancestor).unique() {
-        for (vidx, r) in compute_rewards_for(state, panorama, bhash).enumerate() {
+    let mut prev_block = payout_block;
+    while let Some(prop_hash) = prev_block.parent() {
+        for (vidx, r) in compute_rewards_for(state, panorama, prop_hash).enumerate() {
             match rewards[vidx].checked_add(*r) {
                 Some(sum) => rewards[vidx] = sum,
                 None => {
-                    error!(
-                        "rewards for {:?}, {} + {}, saturate u64",
-                        vidx, rewards[vidx], r
-                    );
+                    let r0 = rewards[vidx];
+                    error!("rewards for {:?}, {} + {}, saturate u64", vidx, r0, r);
```

Comment:
Why are we not panicking here?

Reply:
It feels like in this unlikely case (one would need to have single-handedly finalized 18 million blocks in a single era, which can't happen with our — or any sane — configuration), it's better to just pay out `u64::MAX` instead of crashing the whole network?
Anyway, it's an edge case that should be impossible. Happy to panic, if you prefer.

Reply:
At least for now, while we're still working on this, I think it's better to panic. You're saying that it would equal to finalizing 18kk blocks in a single era by a single validator. This may be true in theory (and hopefully will be in practice as well) but maybe, for some reason, we DO payout `u64::MAX` to a (or all) validator. It would mean that we're doing something wrong – either a bug or a misconfiguration. In either case it's bad and we should fix it. Logs, even `error` level, are too easy to miss.

Reply:
Thanks.